### PR TITLE
use expression operators

### PR DESCRIPTION
**Pull request by yetongumich**

### CODE REVIEW

**Comment by yetongumich** [comment]
ci shall pass after the operator PR in gtsam is merged

File: tests/constrainedExample.h [hunk 1]
```diff
@@ -32,44 +32,27 @@ using gtsam::Vector1;
...
 
 Double_ x1_expr(x1_key);
 Double_ x2_expr(x2_key);
-Double_ cost1_expr(cost1, x1_expr, x2_expr);
-Double_ cost2_expr(cost2, x1_expr, x2_expr);
-Double_ constraint1_expr(constraint1, x1_expr, x2_expr);
+Double_ cost1_expr = x1_expr + ExpExpression(Double_(0.0) - x2_expr);
+Double_ cost2_expr = x1_expr * x1_expr + 2.0 * x2_expr + Double_(1.0);
+Double_ constraint1_expr =
+    x1_expr + x1_expr * x1_expr * x1_expr + x2_expr + x2_expr * x2_expr;
```
Comment (dellaert) [inline tests/constrainedExample.h:55]: Double_ x1(1), x2(2);
blah (x1 + exp(-x2));
blah (pow(x1, 2) + 2*x2 + 1);
blah (x1 + pow(x1, 3) + pow(x2, 2));

Not in a separate file

Comment (yetongumich) [reply tests/constrainedExample.h]: Done

File: tests/constrainedExample.h [hunk 2]
```diff
@@ -32,44 +32,52 @@ using gtsam::Vector1;
...
-  double result = x1 + x1 * x1 * x1 + x2 + x2 * x2;
-  if (H_x1) H_x1->setConstant(1 + 3 * x1 * x1);
-  if (H_x2) H_x2->setConstant(1 + 2 * x2);
-  return result;
+/// Pow functor used for pow function.
+class PowFunctor {
+ private:
+  double c_;
+
+ public:
```
Comment (dellaert) [inline tests/constrainedExample.h:51]: Cool! Should c not be an int, though?

File: tests/constrainedExample.h [hunk 3]
```diff
@@ -32,44 +32,52 @@ using gtsam::Vector1;
...
+
+  double operator()(const double& x,
+                    gtsam::OptionalJacobian<1, 1> H1 = boost::none) const {
+    if (H1) H1->setConstant(c_ * pow(x, c_ - 1));
+    return pow(x, c_);
+  }
+};
+
+/// Pow function.
+Double_ pow(const Double_& x, const double& c) {
```
Comment (dellaert) [inline tests/constrainedExample.h:62]: Same, c int

File: tests/constrainedExample.h [hunk 4]
```diff
@@ -32,44 +32,52 @@ using gtsam::Vector1;
...
+
 Symbol x1_key('x', 1);
 Symbol x2_key('x', 2);
 
-Double_ x1_expr(x1_key);
-Double_ x2_expr(x2_key);
-Double_ cost1_expr(cost1, x1_expr, x2_expr);
-Double_ cost2_expr(cost2, x1_expr, x2_expr);
-Double_ constraint1_expr(constraint1, x1_expr, x2_expr);
+/// Create cost and constraint expressions.
```
Comment (dellaert) [inline tests/constrainedExample.h:76]: Amazing! Now you can move them out of this file and use them in-situ.

**Review by dellaert — APPROVED** [review]

File: tests/constrainedExample.h (hunk 2 above)
Comment (yetongumich) [reply tests/constrainedExample.h:51]: we are only using int in the example, but we can make it double to be general.

File: tests/constrainedExample.h (hunk 4 above)
Comment (yetongumich) [reply tests/constrainedExample.h:76]: As it will appear in multiple files, I still prefer to put it here to avoid duplicate code.

**Pull request merged**